### ISSUE TITLE
System Tray

### ISSUE THREAD

**Issue by KesaiArts01** [godotengine/godot-proposals#2349]
### It would be nice, when exporting the project for Windows, MacOS or Linux, an option that runs the program on the System Tray and define some functions if the user opens the Context Menu on the icon.

Concept:
![godotSystemTray](https://user-images.githubusercontent.com/51276652/109193019-b3200b00-7776-11eb-8a02-fe4f7fd0985b.png)

**Comment by pycbouh** [comment]
Please follow the proposal template and provide all required information for your proposal to be considered.

**Comment by KesaiArts01** [comment]
> Please follow the proposal template and provide all required information for your proposal to be considered.

Ok, updating...

**Comment by Calinou** [comment]
As this feature's implementation is highly OS-specific and not relevant to most games, I would recommend you to use a third-party utility like [RBTray](https://github.com/benbuck/rbtray) instead. It lets you minimize any application to the system tray by right-clicking the Minimize button.

**Comment by pycbouh** [comment]
> > Please follow the proposal template and provide all required information for your proposal to be considered.
> 
> 
> 
> Ok, updating...

Thank you, but I think you've removed all the section titles and as a result you don't have sections and everything is bold.

**Comment by KesaiArts01** [comment]
> > > Please follow the proposal template and provide all required information for your proposal to be considered.
> > 
> > 
> > Ok, updating...
> 
> Thank you, but I think you've removed all the section titles and as a result you don't have sections and everything is bold.

It happens because of `###` on each paragraph. Updating again...

**Comment by takelime** [comment]
any updates on this? This is not "unimportant to most games" but incredibly relevant to many devs out there. No matter if it's for games or not (Godot isn't just good for making games btw)

**Comment by takelime** [comment]
+ Engines like Unity, UE, CryEngine & GameMaker support notifications as well...

**Comment by Calinou** [comment]
> any updates on this? This is not "unimportant to most games" but incredibly relevant to many devs out there. No matter if it's for games or not (Godot isn't just good for making games btw)

As far as I know, nobody has started working on this feature yet.

> Engines like Unity, UE, CryEngine & GameMaker support notifications as well...

Sending desktop notifications is covered by https://github.com/godotengine/godot-proposals/issues/1338.

**Comment by CsloudX** [comment]
I like this feature too, and I wish I can hide taskbar icon, like DestopPet App

**Comment by shrewdlogarithm** [comment]
I'll add my 10p to this - I've recently found myself using Godot to create simple GUI apps and they would benefit from being 'Minimize to Tray' capable

It's also not a 100% Windows thing - when I previously used nwjs it supported this concept on MacOS and some Linux Desktops too - docs here if anyone's interested

https://docs.nwjs.io/en/latest/References/Tray/

**Comment by Bauxitedev** [comment]
If you're using godot-rust you can use this crate to make a tray icon on Win/Mac/Linux: https://github.com/olback/tray-item-rs

**Comment by shrewdlogarithm** [comment]
> If you're using godot-rust you can use this crate to make a tray icon on Win/Mac/Linux: https://github.com/olback/tray-item-rs

Thats interesting and useful to know - thanks!

**Comment by fire** [comment]
https://github.com/zserge/tray Found a single header, super tiny C99 implementation of a system tray icon with a popup menu.on Github.

**Comment by Calinou** [comment]
> [zserge/tray](https://github.com/zserge/tray) Found a single header, super tiny C99 implementation of a system tray icon with a popup menu.on Github.

The upstream repository doesn't appear to be actively maintained, so I'd recommend using the [fork](https://github.com/zserge/tray/issues/20) instead.

**Comment by maxeonyx** [comment]
I'm interested in using this feature for my multiplayer game, for which I've made a separate "dedicated server" GUI scene export, which is something that will tend to run in the background.

**Comment by BadCoyoteStudios** [comment]
I'm writing an app right now that could use this functionality. If Godot is going to be used for applications or advanced games, functionality like this will really help. In the meantime, I may have to move to MAUI.

**Comment by FeralBytes** [comment]
It would be nice if this tiny library could be included in core: https://github.com/dmikushin/tray  and can @Calinou as a contributor to RBTray, can you confirm that is a Windows specific tool? I am hoping to create minimize to tray functionality for a parental tool/Firewall manager on my children's Linux systems.

**Comment by Calinou** [comment]
> and can @Calinou as a contributor to RBTray, can you confirm that is a Windows specific tool?

RBTray is Windows-only, but it's likely feasible to create an alternative that works on macOS and Linux too.

**Comment by OsakiTsukiko** [comment]
this would also be very useful for making launchers for games using godot, once the main game launches the launcher can go to tray and maybe have the ability to add a few options to it like quit, stop, etc

**Comment by Lance-Starkie** [comment]
Yeah, I've got this same issue, for a project I'm working on, would be very nice for a window manager, as playing hot potato whenever the main window gets closed seems like a nightmare to implement. Being able to just retire it to the taskbar is a really nice clean solution there.

**Comment by foxp552** [comment]
> any updates on this? This is not "unimportant to most games" but incredibly relevant to many devs out there. No matter if it's for games or not (Godot isn't just good for making games btw)

Same bro

**Comment by Calinou** [comment]
@foxp552 Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by foxp552** [comment]
Yes sir  |(0_0)\

**Comment by unfa** [comment]
Ok, I love that this works, even if not on Linux (yet).
I wish there was a way to make the application skip taskbar when a status indicator is present.
I am making a productivity application that is supposed to be always on-screen, and having it occupy the taskbar is an unnecessary clutter - keeping it tot he status icon only would be preferable.
Do you consider that a related feature?

**Comment by Calinou** [comment]
> I am making a productivity application that is supposed to be always on-screen, and having it occupy the taskbar is unnecessary clutter - keeping it to the status icon only would be preferable.
> Do you consider that a related feature?

There's a proposal tracking this: https://github.com/godotengine/godot-proposals/issues/8467

**Comment by eumario** [comment]
@Calinou I wanted to contribute to this, but I don't know how I would convert my Python code over into C++ code for Godot.  I was recently working with another Project to implement a StatusIndicator to their app.  The problem being they were using Gtk4, which had dropped StatusIndicator from the library.  So I was able to setup a library, that specifically used no parts of Gtk3, or Qt, though it only uses Gtk4 for GLib.MainLoop() to keep the process running while it is separate from main app.  All through using DBus to do it.  I think something similar can be done with Godot, since I see it being used on the Linuxbsd side for other things.

As for the Icon being file based, if you provide the full path to the icon, (That already exists in the File System) it can be used, and if a custom icon that is included in the project is what is wanted, then simply writing the image to a tmpfile, and providing the full path to the tmpfile over the DBus communication channel to set the icon, should work just fine.

Here's the code I am using to implement the Tray Icon:  https://github.com/eumario/StreamController_TrayIcon/blob/main/backend/trayicon.py

I just don't know how I would be able to implement it in the C++ source code.

**Comment by Delsin-Yu** [comment]
As mentioned in the Contributors Chat that godotengine/godot#89973 implements this proposal.

**Comment by scgm0** [comment]
https://github.com/godotengine/godot/pull/106218 added use for SDL3. I'm wondering if we can also update the existing tray implementation to use SDL3's tray API? Since SDL3's tray is cross-platform and supports Linux, whereas Godot's current tray API lacks Linux implementation.

**Comment by Calinou** [comment]
> I'm wondering if we can also update the existing tray implementation to use SDL3's tray API?

I don't think we should use SDL directly for that, as it would be inconsistent across platforms to use SDL only on Linux. However, we can copy its relevant code since it's under a permissive license.

**Comment by scgm0** [comment]
> > I'm wondering if we can also update the existing tray implementation to use SDL3's tray API?
> 
> I don't think we should use SDL directly for that, as it would be inconsistent across platforms to use SDL only on Linux. However, we can copy its relevant code since it's under a permissive license.

Should SDL's Tray API also support Windows and macOS? It seems possible to migrate the existing tray implementation to SDL.

**Comment by windkb777** [comment]
I really need this code.  I hope the official can add it in stable version.

Minimizing the taskbar window to the tray is just like one line code
or you can change the root window visible flag in the StatusIndicator component.

By the way, Can I Change the root class to the "Node" instead of the "Windows" ?

**Comment by jaysunxiao** [comment]
More and more people are using godot, and many are using it not only for game development but also for application development, hoping to support system tray.

**Comment by Calinou** [comment]
This is already implemented on Windows and macOS: https://docs.godotengine.org/en/latest/tutorials/ui/creating_applications.html#doc-creating-applications-tray-icon

An implementation for Linux is still missing though. https://github.com/godotengine/godot/pull/89973 implements this for Linux, but as https://github.com/godotengine/godot/pull/80211#issuecomment-3014169250 mentions, it's a bit on the hacky side.

**Comment by jaysunxiao** [comment]
This is exactly what I wanted, thanks
